### ISSUE TITLE
Machine working stock pile

### ISSUE THREAD

**Issue by vogonistic** [TeamPorcupine/ProjectPorcupine#1153]
In #1084 we started a tangential discussion on how we handle the stock we are feeding machines. 

When a machine has material it can work with, do we have a visible tile we store it on or is it held inside the machine itself. I can see it working both ways. A hopper makes sense for some machines but deliberate work should be kept hidden unless the job is cancelled. 

I'm worried about something that uses multiple inputs will take up a lot of slots and become huge. 

Factorial keeps inventory buffers internally. Rimworld uses internal buffers for the current job. Last time I played, the only thing with a hopper was the food machine. Can't remember the name now. 

I prefer internal primarily. What do you think?

**Comment by kd7uiy** [comment]
I think that we need at least some kind of an indicator if a good is present, or not present, but I agree that showing on the screen all of the time wouldn't always work. Perhaps we have some system that shows when a resource is getting low, such as a green for fully stocked, yellow for less than a minute supply, and red if we are about to, or have, run out of a material?

**Comment by vogonistic** [comment]
I think it should be visually distinctive when a machine is doing something or not. It might be a good idea to have some kind of visual indicator to differentiate between "low stock, about to stop" and running, but that doesn't feel as important to me. When it stops, that should be obvious.

If we go with the internal buffer and you see a machine that isn't working, a quick click on it tells you that not all slots are full. It requires new UI and some changes to the furniture though.

**Comment by vogonistic** [comment]
Output might be good to put on a visible tile since that is freely available inventory.

**Comment by kd7uiy** [comment]
Let's say instead of a metal smelter we are, say, showing ammo stockpiles in a machine. We might want to know in that case if we are getting low. And the point isn't just to show that a machine stopped, but why it stopped. If there's 5 required resources for a tile, which one of the 5 is low?

**Comment by vogonistic** [comment]
I'm not sure it necessary to always be that specific at a glance.

It would be possible to do similar things to we are doing for power where we show a symbol over the machine. If we do that, we have to figure out how to show that all 5 different resources are missing, or even worse, all 5 resources are missing, it needs O2 to operate and has no power.

Personally, I think knowing that it's off is enough and then you can investigate why if you think it should be running. Otherwise we risk having lots of missing resource symbols all over machines that you haven't told them to work on anyway. There is also another dimension to the problem. If we queue up two different possible jobs on the same machine. Maybe, smelt copper and smelt iron, but both are missing resources. Which one do we show?

**Comment by kd7uiy** [comment]
Fair point, I see where you are coming from. It might still be a good idea to give an indicator that we are running low on supplies, in addition to working well and not working, but I'm certainly okay now with not showing everything.

**Comment by bjubes** [comment]
I prefer the rimworld apporach. When you see an animation of the furniture/character doing the job, its obvious that all the materials are present.

**Comment by sboigelot** [comment]
I also do prefer the rimworld apporach. It's great to see what ingredient are in use for cooking a meal for instance. Also it allow to have an item with the state of the job, and switch to another job by moving the item out of the furniture. I often have multiple character working on one single crafting table with concurrent long running job but at different time frame

**Comment by TomMalbran** [comment]
We could totally do both kinds of machines. Ones with internal storage, others with outside like our metal smelter.

In Factorio there is a view that will show which type of ingredients are inside a machine, and you have a dialog for each machine to show all the details.

Factorio also has a mod to show issues with the machine by just having a dot (big enough to see it clearly but small enough that doesn't bother). A green dot means that everything is fine and the machine is working. A red would mean that there are ingredients missing so the machine can't continue producing. And a yellow dot means that the output is full, so the machine can't produce until there is space to place the output. We can have an option to enable disable the dots, and using a dialog to show what is in the input and in the output you can figure out what is happening.

**Comment by dusho** [comment]
for now I would keep it outside.. clearly visible that resources are brought there (at least while we are still in early development and this will make it easier to find out what is going where and see some inconsistencies at glance)
also, in factorio you can full control over unloading/loading the items in your inventory.. here you have not..  so e.g. you gather some resources at input of machine, but you decided, you don't want any more output of this machine. let's then turn off the machine and items on input will be clearly available for other machines to use (instead of grabbing some items from 'inside')
I'm not worried about having too much input slots.. input slots will be part of furniture size ( so with 2x2 you can have 2 inputs without input obstructing furniture sprite too much), output slots can usually be outside the furniture size..
